[RF] Avoid clearing full RunContext after each NLL evaluation

guitargeek opened the pull request:
In the batched evaluation mode, the RunContext object that cached the values of nodes in the computation graph has so far been completely cleared after each NLL evaluation.

In this PR, a new class `RunContextTracker` is introduced. It makes use of the `RooChangeTracker` class to determine whether a cached result for a given node needs to be recomputed after the change of a given fundamental parameter or not.

This should speed up likelihood fits, as often only one fundamental parameter is changed at the time for the numeric gradient determination.

For the example [rf303_conditional.C](https://root.cern.ch/doc/master/rf303__conditional_8C.html), a reproducible 10 % speedup was observed in batch mode after this PR. The speedup is expected to be more significant for wider computational graphs where more intermediate node results can be kept in the RunContext after each evaluation.

This PR implements one of the ideas suggested in https://github.com/root-project/root/issues/6557, namely the bullet point of "Don't clear all intermediate values in batch fits between fit cycles. Only the ones that changed."

This might be interesting for @lmoneta, @hageboeck and @manolismih.

hageboeck commented on roofit/roofitcore/inc/RooAbsOptTestStatistic.h:
```diff
@@ -91,6 +93,9 @@ class RooAbsOptTestStatistic : public RooAbsTestStatistic {
   Bool_t      _optimized ; //!
   double      _integrateBinsPrecision{-1.}; // Precision for finer sampling of bins.
 
+  bool _batchMode = false; //!
+  bool _originallyCacheOptimized = false;
```
```suggestion
  double      _integrateBinsPrecision{-1.}; // Precision for finer sampling of bins.
  Bool_t      _optimized ; //!

  bool _batchMode = false; //!
  bool _originallyCacheOptimized = false;
```
Pack the bools into one location to reduce padding bytes. Remember that doubles need to be aligned to 8 bytes, so packing anything between 1 to 7 bools in front of them is bad.

Further, does `_originallyCacheOptimized` take part in I/O?
- If yes, you need to increment the class version.
- If no, remove it from I/O using `//!`

guitargeek replied:
I moved the bools to the end.

I now increased the class version and kept both `_batchMode` and `_originallyCacheOptimized` in the IO because the old `_batchedEvaluations` flag in RooNLLVar was also not removed from the IO.

hageboeck replied:
Let's maybe overthink this. The fewer things take part in I/O, the better. Otherwise, evolving the class becomes very cumbersome.
Can the state of those flags be reconstructed when a fit starts? If yes, remove those from I/O.

hageboeck commented on roofit/roofitcore/inc/RooAbsOptTestStatistic.h:
```diff
@@ -55,6 +55,8 @@ class RooAbsOptTestStatistic : public RooAbsTestStatistic {
   Bool_t isSealed() const { return _sealed ; }
   const char* sealNotice() const { return _sealNotice.Data() ; }
 
+  void setBatchMode(bool on = true);
```
Add minimal doxygen comment: `/// <one sentence>.`

hageboeck replied:
Don't forget the `.` at the end for doxygen to know what's the "brief" and what's the extended part.

guitargeek replied:
Done!

hageboeck commented on roofit/roofitcore/src/RooAbsOptTestStatistic.cxx:
```diff
@@ -857,3 +862,27 @@ void RooAbsOptTestStatistic::setUpBinSampling() {
     _funcClone = newPdf;
   }
 }
+
```
Add doxygen comment, preferably using
```
/// Short description >>> . (Dot) <<<
/// A bit more description.
/// \param on <What's that parameter doing.>

guitargeek replied:
Done!

hageboeck commented on roofit/roofitcore/src/RooAbsOptTestStatistic.cxx:
```diff
@@ -859,3 +862,27 @@ void RooAbsOptTestStatistic::setUpBinSampling() {
+    RooArgList serverList{};
+    _funcClone->treeNodeServerList(&serverList);
+    for (auto * arg : serverList) {
+        if(arg->operMode() != Auto) {
+            arg->setOperMode(Auto);
+        }
+    }
```
Your editor is indenting with 3 spaces. Although legacy code is not always consistent in RF, the convention is 2 spaces and *never* tabs.

guitargeek replied:
Done, now it's 2 space.

hageboeck commented on roofit/roofitcore/CMakeLists.txt:
```diff
@@ -227,6 +227,7 @@ ROOT_STANDARD_LIBRARY_PACKAGE(RooFitCore
     RooFitLegacy/RooCatTypeLegacy.h
     RooFitLegacy/RooCategorySharedProperties.h
     RooFitLegacy/RooTreeData.h
+    RunContextTracker.h
```
Maybe move above the legacy block? Not really necessary, though.

guitargeek replied:
Done.

hageboeck commented on roofit/roofitcore/inc/RunContextTracker.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef roofit_roofitcore_RunContextTracker_h
+#define roofit_roofitcore_RunContextTracker_h
+
+#include "RooChangeTracker.h"
+#include "RunContext.h"
```
Replace by forward decls. Both are only used via `*` and `&`.

hageboeck replied:
`RooAbsArg/Real` is also not declared.

guitargeek replied:
Done.

hageboeck commented on roofit/roofitcore/inc/RunContextTracker.h:
```diff
@@ -0,0 +1,43 @@
+
+   // member variables
+
+   std::unordered_map<const RooAbsReal *, std::unique_ptr<RooChangeTracker>> trackers_;
```
Forgot to include the map and `unique_ptr`.

guitargeek replied:
Done

hageboeck commented on roofit/roofitcore/src/RooNLLVar.cxx:
```diff
@@ -95,6 +97,9 @@ RooNLLVar::RooNLLVar(const char *name, const char* title, RooAbsPdf& pdf, RooAbs
   pc.defineInt("extended","Extended",0,kFALSE) ;
   pc.defineInt("BatchMode", "BatchMode", 0, false);
 
+  std::cout << "Hi " << RooCmdConfig::decodeIntOnTheFly("RooNLLVar::RooNLLVar","BatchMode",0,false,arg1,arg2,arg3,arg4,arg5,arg6,arg7,arg8,arg9) << std::endl;
+  std::cout << "Hi " << pc.getInt("BatchMode") << std::endl;
+
```
Remove.

guitargeek replied:
Done.

hageboeck commented on roofit/roofitcore/src/RunContextTracker.cxx:
```diff
@@ -0,0 +1,71 @@
+
+RunContextTracker::RunContextTracker(RooBatchCompute::RunContext const &runContext)
+{
+   for (auto const &item : runContext.ownedMemory) {
+      addTracker(item.first);
+   }
```
That's smart. Start tracking only after first run. 👍

hageboeck commented on roofit/roofitcore/src/RunContextTracker.cxx:
```diff
@@ -0,0 +1,71 @@
+#include "RooMsgService.h"
+
+#include "ROOT/RMakeUnique.hxx"
+
```
Doxygen.

guitargeek replied:
Doxygen comments added for all member functions.

hageboeck commented on roofit/roofitcore/inc/RunContextTracker.h:
```diff
@@ -0,0 +1,43 @@
+
+   // member variables
+
+   std::unordered_map<const RooAbsReal *, std::unique_ptr<RooChangeTracker>> trackers_;
```
RooFit convention would be
```suggestion
   std::unordered_map<const RooAbsReal *, std::unique_ptr<RooChangeTracker>> _trackers;
```

I suggest to not start a third convention inside root ...

guitargeek replied:
Sure!

hageboeck commented on roofit/roofitcore/src/RunContextTracker.cxx:
```diff
@@ -0,0 +1,71 @@
+            runContext.spans.erase(found);
+         }
+      } else {
+         if (runContext.spans.find(arg) != runContext.spans.end()) {
+            oocxcoutD(&caller, FastEvaluations)
+               << "Value of " << arg->GetName() << " kept in RunContext by RunContextTracker::cleanRunContext"
+               << std::endl;
+         }
```
Note that the map lookup will always run, irrespective of whether debug messages are on. I suggest wrapping this in
```c++
  if (RooMsgService::getInstance().msgLvl() == RooFit::DEBUG && <check message topic>) {
```

guitargeek replied:
Done, wrapped in
```C++
if (oodologD(&caller, FastEvaluations)) {    }
```

hageboeck commented on roofit/roofitcore/src/RunContextTracker.cxx:
```diff
@@ -0,0 +1,71 @@
+   RooArgList coll{};
+   absReal->treeNodeServerList(&coll);
+
+   // Should be simplyfied by only using the root node
```
Why not track `absReal` directly? If dirty propagation works, the tracker will do the right thing.

I assume that this is to correctly wipe the server nodes from the `RunContext`, but it looked like the logic in NLLVar would do the same, so almost every tracker would be installed many times.

No matter what you do in the end, the docs (and hopefully the name) of this function should say that it actually installs trackers for *all* daughter nodes. This can become *very* expensive for big models, so be careful to hide this fact from the outside world.

guitargeek replied:
You are right, indeed it's sufficient to track `absReal` directly.

hageboeck commented on roofit/roofitcore/test/testRunContextTracker.cxx:
```diff
@@ -0,0 +1,215 @@
+   };
+
+   // Create the NLL object
+   auto *nll = model.createNLL(dataSet, RooFit::BatchMode(true));
```
Try to keep tests leak free, so one could run valgrind or address sanitizer on them if desired.
```suggestion
   std::unique_ptr<RooNLLVar> nll( model.createNLL(dataSet, RooFit::BatchMode(true)) );
```

guitargeek replied:
Done!

hageboeck reviewed (commented):
Nice to see that this is coming. I added some comments inline.

The title of the second commit should read "Move" instead of "Mode"

guitargeek commented:
Hi @hageboeck, thank you very much for the review! I will fix the typo in the commit message in the end once I `--autosquash` the fixup commits.

I addressed all of your comments, but for one comment I'm not sure if I made the right choice. I'll comment about this inline.

guitargeek commented:
@phsft-bot build

hageboeck commented on roofit/roofitcore/inc/RooAbsOptTestStatistic.h:
```diff
@@ -31,7 +31,7 @@ class RooAbsOptTestStatistic : public RooAbsTestStatistic {
   RooAbsOptTestStatistic() ;
   RooAbsOptTestStatistic(const char *name, const char *title, RooAbsReal& real, RooAbsData& data,
-			 const RooArgSet& projDeps, const char* rangeName=0, const char* addCoefRangeName=0,
+			 const RooArgSet& projDeps, bool batchMode, const char* rangeName=0, const char* addCoefRangeName=0,
```
In general, this will break interfaces for the people out there. If somebody derived their own class from this interface, though, you now broke their code.
*This one time here*, it's maybe OK, because I doubt that many people did actually write their own implementation, and call this constructor.

When you break an interface, the minimum is to add a comment in the release notes. If people advance to 6.26, and their code doesn't compile, they can at least find out that this was intended, and not a bug.

manolismih replied:
It could be the last argument and have a default value so as not to break the interface, yes.

hageboeck commented:
> Hi @hageboeck, thank you very much for the review! I will fix the typo in the commit message in the end once I `--autosquash` the fixup commits.
> 
> I addressed all of your comments, but for one comment I'm not sure if I made the right choice.
> 
> My hesitation concerns the new member variables in `RooAbsOptTestStatistic` that relate to the batched mode: I don't know if they should take part in the IO or not. On one hand, I don't know why flags about how the test statistic is evaluated need to be stored. On the other hand, the `_batchEvaluations` flag that is currently in `RooNLLVar` and that this PR moves to `RooAbsOptTestStatistic` [is not excluded from the IO](https://github.com/root-project/root/blob/master/roofit/roofitcore/inc/RooNLLVar.h#L82). I guess you had a good reason for this choice, so I kept the new flags in `RooAbsOptTestStatistic` also for the IO. You think this is the right choice?

The reason is simple: When `_batchMode` is not persistified, you create a test statistic in BatchMode, write it to file, read it back, it will run in scalar mode. That's at best "annoying".

> 
> Since you mentioned IO: I also increased the class version of `RooNLLVar` since the `_batchEvaluations` flag got moved out of the class.

That's good, but the above will happen also now. If you wrote an old test statistic to a file, which had the batch flag set in the derived class, it will now run in scalar mode, because nobody moves the info up to the base class. It's somehow possible to do that, but I need to look this up myself.
Let's first think if this is the best way to do it, and solve the schema evolution later.

Remind me, please: Why did the flag have to go to the base class?

NB: There's also the RooChi2Var, which could make use of the batch mode. Maybe for this alone, it's already nice to have the flag in the base.

hageboeck commented:
Ups, sorry. Didn't want to close this. 🙂

guitargeek commented:
Hi @hageboeck and @manolismih, thanks for you comments! Sorry for the late answer, it took me some time to come up with a better solution

Making the `RooAbsOptTestStatistic` aware of the batchmode is the biggest challenge in this PR, as you have noticed.
Like Stephan suggested, I have tried no to pass as much information as possible in the constructors to hopefully avoid some of the new member variables I introduced here. In part I have succeeded, because now the cache optimization will now always be skipped in the construction in batch mode is enabled. So I don't need a `_originallyCacheOptimized` flag to remember if the optimization has to be reverted after the construction which was not the nicest hack anyway.

Still, I could not do without a new `_batchMode` flag in `RooAbsTestStatistic` because the cache optimization can also happen after the construction via `RooAbsTestStatistic::constOptimizeTestStatistic`.

Adding a batch mode flag to the constructor also required me to change all the constructor signatures of `RooAbsTestStatistic` and derived classes. I took that opportunity to revisit how the configuration parameters are passed such that if the future it will be much easier to add a new parameter. This has been a significant refactor, so I made a new PR with these changes excluding the `RunContextTracker` which we could introduce in a second step:

https://github.com/root-project/root/pull/7616

guitargeek commented:
I'll close it for now because it depends on https://github.com/root-project/root/pull/7616.

manolismih commented:
After the latest design options finishing this PR is no longer needed! I will copy the necessary code to track the changes in the parameters of the model and call compute only for the nodes affected in the RooFitDriver.

guitargeek commented:
Okay! I will close this PR for now.
